### ISSUE TITLE
Docker daemon hangs and doesn't respond

### ISSUE THREAD

**Issue by rakeshbalguri** [moby/moby#36050]
**Description**
The docker daemon hangs and doesn't respond to any commands. Docker stats shows empty and no stats are shown. All the containers shows the health as "unhealthy"

**We faced this problem many times but we were unable to find the root cause of this problem.**
**We would like to know the root cause and fix it**

**Steps to reproduce the issue:**
1. Start the containers using docker-compose up
2. Wait for couple of days
3. Docker daemon hangs and doesn't respond

**Describe the results you received:**
Docker daemon hangs and doesn't respond
All the containers shows the health as "unhealthy"
**Describe the results you expected:**
Docker daemon shouldn't hang and should respond to all commands
All the containers should show the health as "healthy"

**Additional information you deem important (e.g. issue happens only occasionally):**

**Output of `docker version`:**

```
[root ~]# docker version
Client:
Version:         1.12.6
API version:     1.24
Package version: docker-common-1.12.6-11.el7.centos.x86_64
Go version:      go1.7.4
Git commit:      96d83a5/1.12.6
Built:           Tue Mar  7 09:23:34 2017
OS/Arch:         linux/amd64
 
Server:
Version:         1.12.6
API version:     1.24
Package version: docker-common-1.12.6-11.el7.centos.x86_64
Go version:      go1.7.4
Git commit:      96d83a5/1.12.6
Built:           Tue Mar  7 09:23:34 2017
OS/Arch:         linux/amd64
```

**Output of `docker info`:**

```
[root ~]# docker info
Containers: 34
Running: 32
Paused: 0
Stopped: 2
Images: 70
Server Version: 1.12.6
Storage Driver: devicemapper
Pool Name: docker-253:1-1179653-pool
Pool Blocksize: 65.54 kB
Base Device Size: 10.74 GB
Backing Filesystem: xfs
Data file: /dev/loop2
Metadata file: /dev/loop3
Data Space Used: 8.141 GB
Data Space Total: 107.4 GB
Data Space Available: 99.23 GB
Metadata Space Used: 17.6 MB
Metadata Space Total: 2.147 GB
Metadata Space Available: 2.13 GB
Thin Pool Minimum Free Space: 10.74 GB
Udev Sync Supported: true
Deferred Removal Enabled: false
Deferred Deletion Enabled: false
Deferred Deleted Device Count: 0
Data loop file: /lvm0/docker/devicemapper/devicemapper/data
WARNING: Usage of loopback devices is strongly discouraged for production use. Use `--storage-opt dm.thinpooldev` to specify a custom block storage device.
Metadata loop file: /lvm0/docker/devicemapper/devicemapper/metadata
Library Version: 1.02.135-RHEL7 (2016-11-16)
Logging Driver: journald
Cgroup Driver: systemd
Plugins:
Volume: local
Network: bridge null host overlay
Swarm: inactive
Runtimes: docker-runc runc
Default Runtime: docker-runc
Security Options: seccomp selinux
Kernel Version: 3.10.0-514.el7.x86_64
Operating System: CentOS Linux 7 (Core)
OSType: linux
Architecture: x86_64
Number of Docker Hooks: 2
CPUs: 8
Total Memory: 125.8 GiB
Name: mstr-oi-000.ca.com
ID: 3T77:GDMN:XELE:26CG:TQ2J:WX7Z:4EWZ:DT6P:GVS3:HWG4:HO3I:AFRI
Docker Root Dir: /lvm0/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Insecure Registries:
127.0.0.0/8
Registries: docker.io (secure)
 
```

**Additional environment details [(AWS,]( #url) VirtualBox, physical, etc.):**
This is a CentOS Linux release 7.3.1611 (Core) 

Attached is the output of **kill -SIGUSR1 $(cat /var/run/docker.pid)
[trace.txt](https://github.com/moby/moby/files/1643143/trace.txt)
**

**Comment by cpuguy83** [comment]
This is a very old version of docker that is maintained by RedHat.

Looking through the stack trace it looks like it is stuck waiting on a stream message from containerd.

<details>

```
goroutine 2935906 [select, 277 minutes]:
google.golang.org/grpc/transport.(*Stream).Header(0xc436b8e000, 0x0, 0x0, 0x4f)
	/builddir/build/BUILD/docker-96d83a5ff6ec0eb9bb7b45192c3048fd3aef5e21/vendor/src/google.golang.org/grpc/transport/transport.go:221 +0x1fb
google.golang.org/grpc.recvResponse(0x2225820, 0x226dce8, 0x0, 0x0, 0x0, 0x0, 0x2217960, 0xc4200140a0, 0x222b6e0, 0xc4200140c0, ...)
	/builddir/build/BUILD/docker-96d83a5ff6ec0eb9bb7b45192c3048fd3aef5e21/vendor/src/google.golang.org/grpc/call.go:54 +0x32
google.golang.org/grpc.Invoke(0x2227c20, 0xc420010658, 0x17ed595, 0x11, 0x163a0c0, 0xc4318f99e0, 0x163a160, 0x226dce8, 0xc420478000, 0x0, ...)
	/builddir/build/BUILD/docker-96d83a5ff6ec0eb9bb7b45192c3048fd3aef5e21/vendor/src/google.golang.org/grpc/call.go:178 +0x53e
github.com/docker/containerd/api/grpc/types.(*aPIClient).Signal(0xc4204f0018, 0x2227c20, 0xc420010658, 0xc4318f99e0, 0x0, 0x0, 0x0, 0x150, 0xc421891090, 0xc432417438)
	/builddir/build/BUILD/docker-96d83a5ff6ec0eb9bb7b45192c3048fd3aef5e21/vendor/src/github.com/docker/containerd/api/grpc/types/api.pb.go:945 +0xd2
github.com/docker/docker/libcontainerd.(*client).Signal(0xc420372480, 0xc421ae7f00, 0x40, 0x9, 0x0, 0x0)
	/builddir/build/BUILD/docker-96d83a5ff6ec0eb9bb7b45192c3048fd3aef5e21/_build/src/github.com/docker/docker/libcontainerd/client_linux.go:209 +0x156
github.com/docker/docker/daemon.(*Daemon)

**Comment by rakeshbalguri** [comment]
What is the fix for this problem?  How to resolve this contention?

**Comment by cpuguy83** [comment]
@rakeshbalguri Sounds like a bug in the version of containerd that RedHat has packaged.

**Comment by cpuguy83** [comment]
If you aren't tied to the RedHat packaging, I would recommend installing docker-ce from Docker's official repos.

**Comment by rakeshbalguri** [comment]
We are not using any RedHat packaging. We are using Docker which gets shipped by default with CentOS.

**Comment by cpuguy83** [comment]
@rakeshbalguri That is Redhat's fork of Docker, which is both very old and has significant divergence from this codebase.

Here's the official docs for installing Docker on CentOS: https://docs.docker.com/engine/installation/linux/docker-ce/centos/

**Comment by rakeshbalguri** [comment]
Thanks @cpuguy83. What is the immediate remediation until we upgrade docker to the latest version?

**Comment by cpuguy83** [comment]
@rakeshbalguri Need to restart docker
